src/libMessage/Messenger.cpp

```diff
@@ -111,6 +115,227 @@ void NumberToArray(const T& number, vector<unsigned char>& dst,
+                                      account.GetStorageRoot().asArray())
+                               << " Actual: "
+                               << DataConversion::charArrToHexStr(
+                                      tmpStorageRoot.asArray()));
+      return false;
+    }
+  }
+
+  return true;
+}
+
+void AccountDeltaToProtobuf(const Account* oldAccount,
+                            const Account& newAccount,
+                            ProtoAccount& protoAccount) {
+  Account acc(0, 0);
+
+  bool fullCopy = false;
+
+  if (oldAccount == nullptr) {
+    oldAccount = &acc;
+    fullCopy = true;
+  }
+
+  int256_t balanceDelta =
+      int256_t(newAccount.GetBalance()) - int256_t(oldAccount->GetBalance());
+  protoAccount.set_numbersign(balanceDelta > 0);
+
+  uint256_t balanceDeltaNum(abs(balanceDelta));
+  NumberToProtobufByteArray<uint256_t, UINT256_SIZE>(
+      balanceDeltaNum, *protoAccount.mutable_balance());
+
+  uint256_t nonceDelta = newAccount.GetNonce() - oldAccount->GetNonce();
+  NumberToProtobufByteArray<uint256_t, UINT256_SIZE>(
+      nonceDelta, *protoAccount.mutable_nonce());
+
+  if (!newAccount.GetCode().empty()) {
+    if (fullCopy) {
+      protoAccount.set_code(newAccount.GetCode().data(),
+                            newAccount.GetCode().size());
+      protoAccount.set_initdata(newAccount.GetInitData().data(),
+                                newAccount.GetInitData().size());
+      protoAccount.set_createblocknum(newAccount.GetCreateBlockNum());
+    }
+
+    if (newAccount.GetStorageRoot() != oldAccount->GetStorageRoot()) {
+      protoAccount.set_storageroot(newAccount.GetStorageRoot().data(),
+                                   newAccount.GetStorageRoot().size);
+
+      for (const auto& keyHash : newAccount.GetStorageKeyHashes()) {
+        string rlpStr = newAccount.GetRawStorage(keyHash);
+        if (rlpStr != oldAccount->GetRawStorage(keyHash)) {
+          ProtoAccount::StorageData* entry = protoAccount.add_storage();
+          entry->set_keyhash(keyHash.data(), keyHash.size);
+          entry->set_data(rlpStr);
+        }
+      }
+    }
+  }
+}
+
+bool ProtobufToAccountDelta(const ProtoAccount& protoAccount, Account& account,
+                            const bool fullCopy) {
+  uint256_t tmpNumber;
+
+  ProtobufByteArrayToNumber<uint256_t, UINT256_SIZE>(protoAccount.balance(),
+                                                     tmpNumber);
+
+  int balanceDelta =
+      protoAccount.numbersign() ? (int)tmpNumber : 0 - (int)tmpNumber;
+  account.ChangeBalance(balanceDelta);
+
+  ProtobufByteArrayToNumber<uint256_t, UINT256_SIZE>(protoAccount.nonce(),
+                                                     tmpNumber);
+  account.IncreaseNonceBy(tmpNumber);
+
+  if (protoAccount.code().size() > 0) {
+    bool doInitContract = false;
+
+    if (fullCopy) {
+      vector<unsigned char> tmpVec;
+      tmpVec.resize(protoAccount.code().size());
```

Comment:
should check that size() is not excessively large before doing resizing

Reply:
ok